### ISSUE TITLE
Tidal plugin 1.1.0

### ISSUE THREAD

**Issue by bkmill** [michaelherger/lms-plugin-tidal#73]
Just installed the update and now i'm missing all my Tidal Playlists.

**Comment by michaelherger** [comment]
"your playlists": playlists you have created, or playlists you have added to your favorites?

**Comment by bkmill** [comment]
Playlists i have created in Tidal. They still showed in version 1.0.0 and are also still available in Tidal.

**Comment by bkmill** [comment]
Not really, most of them are containing some 300 songs. The playlists are still visible in the Tidal app on my computer and phone. With version 1.0.0 of the new Tidal Plugin everything worked fine.

**Comment by philippe44** [comment]
Are they public or private?

**Comment by bkmill** [comment]
Most of them are private, but i just changed a couple of them to public (using the desktop Tidal app) and that did not change the visibility in LMS. Restarting PCP and/or LMS did make a difference either.

**Comment by bkmill** [comment]
I meant to say: did NOT make a difference. Sorry

**Comment by philippe44** [comment]
I tried again with my playlist: 2 privates and an existing public one that I added to my favorites and they all appear under "Playlists"... Is there anything more you can tell about how you created your playlists?

**Comment by philippe44** [comment]
Can you erase your LMS cache to be sure? (stop LMS and delete /var/lib/squeezeboxserver/cache/cache.db)

**Comment by michaelherger** [comment]
That's pretty much the issue I've tried to address in the past. I don't remember whether there was a well defined rule as to which playlists would show up and which one wouldn't... I fear it's time to re-introduce that not so elegant code...

**Comment by philippe44** [comment]
I would still like to go to the bottom of this as there has to be a logical reason and I know there is another api call as well.

**Comment by philippe44** [comment]
Just to elaborate, there is an API `playlistsAndFavoritePlaylists` but AFAIK, any playlist that I have created is *always* in Favorites and I've not see an option for it not to be. To the OP, have you done something special for that?

**Comment by bkmill** [comment]
Just deleted /mnt/mmcblk0p2/tce/slimserver/Cache/cache.db 
That did not resolve the issue.

**Comment by philippe44** [comment]
And if you create a fresh playlist, does it appear? Are you aware of any settings your would have change wrt blocking your playlists to be in favorites?

**Comment by bkmill** [comment]
I created a new public playlist in the Tidal app. It does NOT show up in LMS.

**Comment by bkmill** [comment]
Yes i have. Some of my playlist are created using the Tidal API and when i create the playlist i don’t add them to the Favorites because what’s the point of having a playlist and adding them to favorites.

But creating a new playlist in the Tidal app does not show up in LMS either.

**Comment by bkmill** [comment]
I just did a small test. I created a playlist using the Tidal API and also added it to my Tidal Favorites. The newly created playlist is now visible in LMS.

**Comment by philippe44** [comment]
Ok, this is really insane - I've created a playlist 20 times using the tidal and it *always* appeared in favorites immediately. Anyway, I will use the other api but I'll have also to think about changes detection

**Comment by philippe44** [comment]
Can you share the code you used to create the playlist or at least the exact api calls?

**Comment by bkmill** [comment]
Yes, of course.

This is the Python code i changed this afternoon to add the playlist:

def createTidalPlaylist(session, playlistName, playlistDescription="", verbose=False):

    user = Session.get_user(session, session.user.id)
    playlistDescription = util_helper.printCurrentDateTime("Created by BarkingMill on", "%d-%m-%Y", False)

    playlist = tidalapi.user.LoggedInUser(session, user.id).create_playlist(playlistName, playlistDescription)

    # TW: 6-4-2024 Uncommented this line
    # TW: No real need to add the playlist to favorites
    # Don't forget to add the newly created playlist to your favorites
    tidalapi.Favorites(session, user.id).add_playlist(playlist.id)

    if playlist is not None:
        if verbose:
            print(f"{util_helper.bcolors.OKGREEN}Playlist {playlist.name}: {util_helper.bcolors.FAIL}{playlist.id}{util_helper.bcolors.OKGREEN} created.{util_helper.bcolors.ENDC}")

        return playlist

    else:
        return None

**Comment by philippe44** [comment]
What python frontend lib do you use? tamland/python-tifal?

**Comment by bkmill** [comment]
https://tidalapi.netlify.app

**Comment by chaseadam** [comment]
It is conceivable that the "normal" UIs automatically add your playlist to your "favorites" and that is you do it with the API, it won't show up in your favorites API call. I was recently playing with the python tidalapi and found user.playlists(), but noted a few playlists missing and then found user.favorites.playlists() and it contained all of the playlists I wanted.

**Comment by philippe44** [comment]
Yes, likely this is the case. I'll change that to reorganize a bit the current version as there is a dedicated api call to get everything.

**Comment by philippe44** [comment]
I'm about to push and updated version, but what is interesting to see is that, at least on my setup, such playlists that are user-defined but removed from favorites through API **disappear from the TIDAL app**. They now remain on LMS and I assume on other apps using the API, but are gone on TIDAL. AFAIR, the TIDAL webapp uses a v2 version of the API for playlist, I traced that recently. So, it might be that, as in Deezer, when v2 is deployed, user playlist must be in favorites.

**Comment by philippe44** [comment]
> I just did a small test. I created a playlist using the Tidal API and also added it to my Tidal Favorites. The newly created playlist is now visible in LMS.
>
And that has never been my experience. Even with 1.1, evert playlist I create in TIDAL app appeared in LMS. And I've tried *many* times.

**Comment by philippe44** [comment]
> I created a new public playlist in the Tidal app. It does NOT show up in LMS.
>
And that never been my experience. Every time I create a playlist in TIDAL, with v1.1 on this plugin, it always appeared in LMS

**Comment by michaelherger** [comment]
Thanks for looking into this, @philippe44!

**Comment by bkmill** [comment]
Thanks very much. Just installed the update and voila, all the playlists are back.

**Comment by philippe44** [comment]
Cool - it's just so weird why, using their official app, I can'reproduce the issue and have a user playlist that is NOT in favorite / collection.

**Comment by bkmill** [comment]
I'm sorry, i think i was a bit premature. Although my "old" Tidal playlists are now visible in LMS, newly created playlists, created by code but also created with the Tidal Desktop app, are NOT visible in LMS. The "new" playlists however are visible in both the Tidal desktop and mobile app. I'm not sure what's wrong.

**Comment by philippe44** [comment]
What country are you in? That does not make sense, I've tried times and times and times.

**Comment by bkmill** [comment]
Netherlands

**Comment by bkmill** [comment]
Yep, i created a playlist, added titles to it in the Tidal desktop app. The playlist is visible in the Tidal app but not in LMS.

**Comment by philippe44** [comment]
@michaelherger : can you try on your side as well? This will drive me insane. I just tried again creating one in the desktop app (Windows) and the playlist was immediately visible in LMS. I did nothing special.

**Comment by bkmill** [comment]
About 60. If needed for testing i can delete them because most of them are generated.

**Comment by philippe44** [comment]
https://github.com/michaelherger/lms-plugin-tidal/assets/9712599/99a33a43-b27e-449a-bd15-c3f7e14064b0

**Comment by bkmill** [comment]
My scenario is the same, the only difference is the playlist not showing up in LMS :-(

**Comment by philippe44** [comment]
I found the issue. It's a paging problem. You have 60 playlist and paging is 50. When I set paging to 2, I only see 2 playlists...

**Comment by bkmill** [comment]
Yep... you're right. I deleted most of my playlists and de newly created ones showed up. Pfieuuuwwww.

**Comment by philippe44** [comment]
Now I need to talk to @michaelherger / spend more time because I was reading the paging mechanism and now I'm totally confused.

**Comment by bkmill** [comment]
Take your time, i'm not in a hurry. Thank you for your effort.

**Comment by michaelherger** [comment]
I think I've found something: I passed `$acb` as the callback for the `_call` method, which would return the result and the response object as the second argument. But `$acb` would consider a second argument an error. Thus would cut short and return only the first page of results, assuming the others failed.

By explicitly calling `$acb->(shift)` we only pass the result set. Which allows `achain` to evaluate all result sets of all the calls.

**Comment by philippe44** [comment]
Yes, my slow brain was getting there as well. Don't you think as well that _nocache => 1 should be set to every amap action?

**Comment by philippe44** [comment]
I've checked and we must not cache the sub-page calls or at least we should follow what the initial request wrt cache was. I think it's not useful to cache these interim results knowing that we want to cache the final result, but more important, if we cache them (say we cache 0..49,50..99 and 100..120) then next time we have a request for the whole lot asking us to not cache or to refresh, then the first 0..49 will be refreshed/not use cache but the others will be taken from the previous cached data, which is wrong. Hence I think the best is to simply not cache sub-pages queries, rather than passing through _nocache/refresh.
